src/main/clojure/jarl/builtins/jwt.clj

```diff
@@ -0,0 +1,308 @@
+        (throw (errors/builtin-ex "failed to parse a PEM certificate"))))))
+
+(defn key-resolver-name [^String certificate]
+  (if (str/starts-with?     certificate "{")
+    (if (str/includes?      certificate "keys") "jwks" "jwk")
+    (if (str/starts-with?   certificate "-----BEGIN PUBLIC KEY-----")
+      "pem"
+      (if (str/starts-with? certificate "-----BEGIN CERTIFICATE-----")
+        "x509"
+        (throw (errors/builtin-ex "failed to extract a Key from the PEM certificate"))))))
+
+(defn key-resolver ^VerificationKeyResolver [certificate alg]
+  (condp = (key-resolver-name certificate)
+    "jwks" (jwks-key-resolver certificate)
+    "jwk"  (jwk-key-resolver  certificate)
+    "pem"  (pem-key-resolver  certificate alg)
+    "x509" (x509-key-resolver certificate)
+    (throw (errors/builtin-ex "no resolver found for %s" certificate))))
+
+(defn verify-hmac [jwt ^String secret]
+  (let [jws (doto (JsonWebSignature.)
+                  (.setDoKeyValidation false)
+                  (.setCompactSerialization jwt)
+                  (.setKey (HmacKey. (.getBytes secret "UTF-8"))))]
+    (.verifySignature jws)))
+
+(defn- verify-no-jwe [header]
+  (if (contains? header "enc")
+    (throw (errors/builtin-ex "JWT is a JWE object, which is not supported"))
+    header))
+
+(defn- strip-quotes [s]
+  (cond-> s (and (str/starts-with? s "\"") (str/ends-with? s "\"")) (subs 1 (dec (count s)))))
+
+(defn decode [jwt]
+  (let [parts       (split-jwt jwt)
+        header      (-> (first  parts) (base64-decode-bytes "header") (String.) json/read-str verify-no-jwe)
+        payload-str (-> (second parts) (base64-decode-bytes "payload") (String.))
+        sig         (-> (last   parts) (base64-decode-bytes "signature") hex/encode-bytes)]
+    (if (= "JWT" (get header "cty"))
+      (decode (strip-quotes payload-str)) ; nested JWT
+      [header (json/read-str payload-str) sig])))
+
+(defn verify-sig [jwt certificate alg]
+  (let [parts  (split-jwt jwt)
+        header (-> (first  parts) (base64-decode-bytes "header") (String.) json/read-str verify-no-jwe)]
+    (if-not (= alg (alg-str->alg (get header "alg")))
+      false
+      (if (hs-alg? alg)
+        (verify-hmac jwt certificate)
+        (let [resolver (key-resolver certificate alg)
+              jws (doto (JsonWebSignature.) (.setCompactSerialization jwt))]
+          (try
+            (.verifySignature (doto jws (.setKey (.resolveKey resolver jws (Collections/emptyList)))))
+            (catch UnresolvableKeyException _ false)))))))
+
+; Since not only the contents/data of a JWT decides the signature, but also the JSON _formatting_,
+; we need to ensure that the JSON produced here looks the same as that produced by OPA.
+; We should probably replace this with something more robust..
+(defn opa-json-fmt [json-str]
+  (-> json-str
+      (str/replace #"(\":)(\S)" "$1 $2")   ; add whitespace after colon
+      (str/replace #"(,)(\")"   "$1 $2"))) ; add whitespace after comma
```

Comment:
This looks a bit brittle, considering that the JWT payload document could have string values containing `:` and `,`. Which is content that we definitely shouldn't modify.

Reply:
Oh, it's brittle for sure, but not quite _that_ brittle — you'll need to match not just `:` or `,` but `":` or `,"` respectively. Obviously not impossible to encounter, but definitely not as likely. The alternative here would be to throw out the JSON library in favor of something that allows that level of fine tuning.. and while that's probably the way to go at some point, I left it with a big disclaimer for now. FWIW, using OPA to _sign_ tokens is pretty much never done outside of unit testing, while verification is obviously everywhere.